### PR TITLE
Fused all-gather+GEMM HBM-buffer kernel for iris.ops

### PR DESCRIPTION
**Pull request by neoblizz**
Adds `all_gather_matmul_hbm_buffer`: a fused kernel that pipelines all-gather and GEMM by splitting workgroups into dedicated fetchers and GEMM workers. Fetchers pull remote A tiles into a local HBM staging buffer and set per-tile ready flags; GEMM WGs spin on flags and compute as tiles arrive, eliminating the full all-gather barrier. Delivers **2.7–3.4× lower latency** vs the barrier-based baseline on 8× MI325X.

## New kernel
- **`iris/ops/all_gather_matmul_hbm_buffer.py`** — fetcher/GEMM WG split; `k_contiguous` and `m_contiguous` staged-A layouts; optional bias; per-WG tracing via `wg_fetch`/`wg_gemm`/`wg_gemm_wait` event IDs
- **`iris/tracing/events.py`** — trace event IDs for per-workgroup profiling

## API / config changes
- **`iris/x/gather.py`** — `hint` vectorization parameter forwarded to `_translate()`
- **`iris/ops/__init__.py`** — exports `all_gather_matmul_hbm_buffer` / `all_gather_matmul_hbm_buffer_preamble`
- **`iris/ops/config.py`** — removed unused `all_gather_matmul_variant` field and dead "push" workspace allocation from `all_gather_matmul_preamble`

## Benchmark & tests
- **`benchmark/ops/bench_all_gather_matmul.py`** — merged baseline and HBM-buffer variants under `@bench.axis("algorithm", ["baseline", "hbm_buffer"])`; `bench_all_gather_matmul_hbm_buffer.py` deleted
- **`tests/ops/test_all_gather_matmul.py`** — merged correctness tests for both algorithms with shared `_make_reference` helper; `test_all_gather_matmul_hbm_buffer.py` deleted

## Results (8× AMD MI325X, float16, N=3584, K=8192)

| Ranks | MxNxK | Baseline (ms) | HBM Buffer (ms) | Speedup | TFLOPS |
|-------|-------|---------------|-----------------|---------|--------|
| 2 | 1024×3584×8192 | 1.67 | 0.78 | **2.1×** | 77 |
| 2 | 16384×3584×8192 | 27.8 | 8.2 | **3.4×** | 117 |
| 4 | 16384×3584×8192 | 27.3 | 8.6 | **3.2×** | 112 |
| 8 | 16384×3584×8192 | 24.4 | 8.9 | **2.7×** | 108 |

TFLOPS
Latency

### CODE REVIEW

**Comment by ryanswann-amd** [comment]
@copilot Can you add proper vectorization hints using the new constexpr method and tests for the hbm buffered all gather matmul?

**Comment by mawad-amd** [comment]
@copilot can you cleanup this PR?

**Comment by mawad-amd** [comment]
@copilot lets sync with main and resolve the conflict. Also remove any unneeded scripts. Minimize the number of files this PR change.

**Comment by mawad-amd** [comment]
@copilot remove any helper scripts that is not the algorithmic improvement or new algorithms. Port benchmarks to iris.bench.

**Comment by mawad-amd** [comment]
@copilot replace `shmem.` with `ctx.` and run the benchmark and comment with a chart please.

**Comment by mawad-amd** [comment]
@copilot Run benchmarks comparing the new `all_gather_matmul_hbm_buffer` against the existing `all_gather_matmul` baseline. Use the `iris.bench` benchmarks at 2, 4, and 8 ranks:

```bash
# Baseline
torchrun --nproc_per_node=2 benchmark/ops/bench_all_gather_matmul.py
torchrun --nproc_per_node=4 benchmark/ops/bench_all_gather_matmul.py
torchrun --nproc_per_node=8 benchmark/ops/bench_all_gather_matmul.py

# New HBM-buffered kernel
torchrun --nproc_per_node=2 benchmark/ops/bench_all_gather_matmul_hbm_buffer.py
torchrun --nproc_per_node=4 benchmark/ops/bench_all_gather_matmul_hbm_buffer.py
torchrun --nproc_per_node=8 benchmark/ops/bench_all_gather_matmul_hbm_buffer.py
```

Post the results here as:
1. **Markdown tables** with columns: M, N, K, baseline latency (ms), HBM-buffer latency (ms), speedup (x) — one table per rank count
2. **Figures** (matplotlib plots uploaded as images) comparing throughput (TFLOPS) across problem sizes for each rank count

If a benchmark fails, fix it and re-run.

**Comment by mawad-amd** [comment]
@copilot Clean up the duplicated code in this PR:

**1. Merge benchmarks into one file.** `bench_all_gather_matmul.py` and `bench_all_gather_matmul_hbm_buffer.py` are nearly identical. Merge them into a single `bench_all_gather_matmul.py` with an `@bench.axis("algorithm", ["baseline", "hbm_buffer"])` that selects the variant. Delete `bench_all_gather_matmul_hbm_buffer.py`.

**2. Merge tests into one file.** `test_all_gather_matmul.py` and `test_all_gather_matmul_hbm_buffer.py` share the same boilerplate (heap init, seeded tensors, torch reference, tolerance check). Merge into one test file parametrized by algorithm. Extract shared setup into a fixture or helper. Delete `test_all_gather_matmul_hbm_buffer.py`.

**3. Remove dead code.** The `"push"` variant workspace allocation in `all_gather_matmul.py` preamble (lines allocating `a_inbox` and `signal_flags`) is unused — no push kernel exists. Remove it. Also remove `all_gather_matmul_variant` from `FusedConfig` if it's not actually used to dispatch between variants.

**4. Re-run benchmarks after cleanup.** Run the merged benchmark at 2, 4, and 8 ranks and post results as:
- Markdown tables with columns: M, N, K, baseline latency (ms), HBM-buffer latency (ms), speedup (x)
- Figures (matplotlib) with **MxNxK on the x-axis** (not just M) comparing TFLOPS across problem sizes for each rank count

Confirm the results match the previous run.

**Comment by ryanswann-amd** [comment]
@copilot+claude-opus-4.6 Please update the buffered all-gather matmul to use the following default parameters, then re-run the sweep:

| Parameter              | Value |
|------------------------|-------|
| `k_per_flag`           | 16    |
| `num_fetch_sms`        | 32    |
| `num_warps`            | 8     |
| `num_stages`           | 3     |
| `first_stage_fetch_sms`| 256   |
| `block_m`              | 128   |
| `block_n`              | 256   |
| `block_k`              | 64    |

Run the merged benchmark at 2, 4, and 8 ranks and post results as:

1. **Markdown tables** with columns: M, N, K, baseline latency (ms), HBM-buffer latency (ms, previous defaults), HBM-buffer latency (ms, new defaults), speedup vs baseline (×). Include both parameter sets for reference.
2. **Figures** (matplotlib) with M×N×K on the x-axis comparing TFLOPS across problem sizes for each rank count, with separate lines for baseline, previous defaults, and new defaults.

**Comment by ryanswann-amd** [comment]
@copilot+claude-opus-4.6 Cherry-pick the config files from my fork: [`feature/auto-config-xops-perf`](https://github.com/ryanswann-amd/iris/tree/feature/auto-config-xops-perf)

Bring in the changes for the auto config only. The major changes are in :
- `benchmark/ops/all_gather_matmul/auto_config.py`
- `benchmark/ops/all_gather_matmul/configs/` (all JSON files under `mi300x/` and `mi355x/`)
- `benchmark/ops/bench_all_gather_matmul.py`

Don't touch `iris/ops/__init__.py` or delete `tests/ops/test_auto_config.py`.

After porting, run benchmark/ops/bench_all_gather_matmul.py on MI355X and MI300X for all sizes which configs were added for. Produce a bar plot for each GPU showing `iris_ms` vs `pytorch_ms` vs `expected_iris_ms` (from the config JSON) for every tuned shape. Flag any shape where measured `iris_ms` drifts more than ~15% from expected in your response.

**Comment by ryanswann-amd** [comment]
@copilot+claude-opus-4.6 You didn't produce a bar plot, nor did you run on 350.

**Comment by mawad-amd** [comment]
@copilot can you please address Ryan's comment.

File: iris/ops/all_gather_matmul_hbm_buffer.py [hunk 1]
```diff
@@ -0,0 +1,718 @@
...
+                        rk = tl.max_contiguous(tl.multiple_of(rk, BLOCK_SIZE_K), BLOCK_SIZE_K)
+                        staged_ptrs = staged_a + rm.to(tl.int64)[:, None] * stride_sa_m + rk[None, :] * stride_sa_k
+
+                        for compile_rank in range(world_size):
+                            if src_rank_idx == compile_rank:
+                                a_tile = iris.x.gather(k_tile, src_view, compile_rank, ctx, hint=(1, BLOCK_SIZE_K))
+                                tl.store(staged_ptrs, a_tile, cache_modifier=".cg")
+
+                    flag_idx = m_tile * NUM_FLAG_GROUPS_K + k_flag_group
+                    tl.atomic_xchg(flags_ptr + flag_idx, 1, sem="release", scope="gpu")
```
Comment (dsidler) [inline iris/ops/all_gather_matmul_hbm_buffer.py:316]: you need a    `tl.debug_barrier()` before the atomic.xchg since the atomic is per wave but the store per block.
in my runs i got validation failures without the barrier

File: benchmark/ops/bench_all_gather_matmul.py [hunk 2]
```diff
@@ -2,11 +2,55 @@
 # SPDX-License-Identifier: MIT
 # Copyright (c) 2026 Advanced Micro Devices, Inc. All rights reserved.
 
-"""Benchmark for fused all-gather + GEMM (iris.ops)."""
+"""Benchmark for all-gather + GEMM: RCCL baseline vs iris HBM-buffer prefetch.
```
Comment (dsidler) [inline benchmark/ops/bench_all_gather_matmul.py:5]: any reason to no longer include validation logic?

Comment (dsidler) [reply benchmark/ops/bench_all_gather_matmul.py:5]: the benchmark code doesn't match the reference calculation in the test:
it passes because we do `torch.ones` here (which also gives an significant perf boost to pytorch)
```
    torch.manual_seed(42 + rank)
    A_sharded = torch.randn(M, K_local, dtype=dtype, device=device)

    torch.manual_seed(123)
    B = torch.randn(K, N, dtype=dtype, device=device)

    A_gathered_list = [torch.zeros(M, K_local, dtype=dtype, device=device) for _ in range(world_size)]
    dist.all_gather(A_gathered_list, A_sharded)
    A_gathered_ref = torch.cat(A_gathered_list, dim=1)
    ref_output = torch.matmul(A_gathered_ref, B)
```

Comment (mawad-amd) [reply benchmark/ops/bench_all_gather_matmul.py:5]: We want to avoid mixing examples, benchmarks and tests. I think tests should live on their own scripts and we only benchmark once we have a correct and we are confident of the implementation. 

@copilot could you please test what David is suggesting here? Can you show us perf before and after applying the suggested fix.

Comment (mawad-amd) [reply benchmark/ops/bench_all_gather_matmul.py:5]: @copilot Ryan suggested some specific shapes and tiling parameters earlier. Can you reproduce the bar chart and reply here please?

Comment (dsidler) [reply benchmark/ops/bench_all_gather_matmul.py:5]: @copilot  check commit c52e5edb2f58d733f2437fd776ada4d87ac4bcda for the pytorch-rccl bench. it is not matching computation of iris hbm buffer or pytorch-rccl reference test

Comment (mawad-amd) [reply benchmark/ops/bench_all_gather_matmul.py:5]: @copilot please address David's request.

**Comment by mawad-amd** [comment]
@copilot remove the charts you uploaded and get this PR ready for final review.

**Comment by ryanswann-amd** [comment]
@copilot resolve the merge conflicts in this pull request

**Comment by ryanswann-amd** [comment]
@copilot can you expand test coverage for the kernels added in this PR?

**Comment by ryanswann-amd** [comment]
@copilot+claude-opus-4.6 I identified that the linked diff fixes the CI changes. Can you cherry pick minimal changes to fix the CI error: https://github.com/ryanswann-amd/iris/compare/39c213daa7b0...5f30678c

**Review by mawad-amd — APPROVED** [review]
